packages/protocol/contracts/governance/LockedGold.sol

```diff
@@ -264,6 +329,199 @@ contract LockedGold is
+      "revoked percents can be only between 1%..100%"
+    );
+
+    address delegatorAddress = getAccounts().voteSignerToAccount(msg.sender);
+    Delegated storage delegated = delegatorInfo[delegatorAddress];
+    require(
+      delegated.totalDelegatedCeloInPercents >= percentageToRevoke,
+      "Not enough total delegated percents"
+    );
+
+    address delegateeAccount = getAccounts().voteSignerToAccount(delegatee);
+    updateDelegatedAmount(msg.sender, delegatee);
+
+    DelegatedInfo storage currentDelegateeInfo = delegated
+      .delegateesWithPercentagesAndAmount[delegateeAccount];
+
+    require(currentDelegateeInfo.percentage >= percentageToRevoke, "Not enough delegated percents");
+
+    currentDelegateeInfo.percentage = currentDelegateeInfo.percentage.sub(percentageToRevoke);
+
+    uint256 delegateeTotalVotingPower = getAccountTotalGovernanceVotingPower(delegateeAccount);
+    uint256 totalReferendumVotes = getGovernance().getAmountOfGoldUsedForVoting(delegateeAccount);
+
+    uint256 totalLockedGold = getAccountTotalLockedGold(delegatorAddress);
+
+    // in a worst case we will remove whole amount of delegator (even when amount is not updated)
+    uint256 amountToRevoke = currentDelegateeInfo.percentage == 0
+      ? currentDelegateeInfo.currentAmount
+      : Math.min(
+        FixidityLib
+          .newFixed(totalLockedGold)
+          .multiply(FixidityLib.newFixedFraction(percentageToRevoke, 100))
+          .fromFixed(),
+        currentDelegateeInfo.currentAmount
+      );
+
+    uint256 unusedReferendumVotes = delegateeTotalVotingPower.sub(totalReferendumVotes);
+    if (unusedReferendumVotes < amountToRevoke) {
+      getGovernance().removeVotesWhenRevokingDelegatedVotes(
+        delegateeAccount,
+        delegateeTotalVotingPower.sub(amountToRevoke)
+      );
+    }
+
+    delegated.totalDelegatedCeloInPercents = delegated.totalDelegatedCeloInPercents.sub(
+      percentageToRevoke
+    );
+
+    currentDelegateeInfo.currentAmount = currentDelegateeInfo.currentAmount.sub(amountToRevoke);
+    totalDelegatedCelo[delegateeAccount] = totalDelegatedCelo[delegateeAccount].sub(amountToRevoke);
+
+    if (currentDelegateeInfo.currentAmount == 0) {
+      delegated.delegatees.remove(delegateeAccount);
+    }
+
+    emit CeloDelegatedRevoked(
+      delegatorAddress,
+      delegateeAccount,
+      percentageToRevoke,
+      amountToRevoke
+    );
+  }
+
+  function updateDelegatedAmount(address delegator) public {
+    address delegatorAccount = getAccounts().voteSignerToAccount(delegator);
+    EnumerableSet.AddressSet storage delegatees = delegatorInfo[delegatorAccount].delegatees;
+    for (uint256 i = 0; i < delegatees.length(); i = i.add(1)) {
+      updateDelegatedAmount(delegatorAccount, delegatees.get(i));
+    }
+  }
+
+  /**
+   * Updates real delegated amount to delegatee.
+   * There might be discrepancy because of validator rewards or extra locked gold.
```

Comment:
The discrepancy will always be that the delegated voting power is always smaller or equal than actual delegated voting power? In that case I'll clarify that in the comment.